In [ ]:
!pip install PyDub
!pip install SpeechRecognition

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 32.8/32.8 MB 32.0 MB/s eta 0:00:00


In [ ]:
import speech_recognition as sr

In [ ]:
from pydub import AudioSegment


Converting audio to the right format

In [ ]:
def convert_to_wav(filename):
  """ Takes an audio file not of the mav type and converts it to mav"""
  #importing audio
  audio = AudioSegment.from_file(filename)
  #new file name
  new_filename = filename.split('.')[0]
  new_filename = new_filename.split('/')[-1] + '.wav'
  #export file as wav
  audio.export(new_filename,format = 'wav')
  print(f"Converting {filename} to {new_filename}...")
  #this project was coded in google colab therefore the file exported have the
  #path as describe bellow, the function will return the path.
  x = '/content/'+new_filename
  return x

In [ ]:
#testing the function
call_1 = '/content/drive/MyDrive/dataset/call_1_stereo.mp3'
call_1_wav = convert_to_wav(call_1)


Converting /content/drive/MyDrive/dataset/call_1_stereo.mp3 to call_1_stereo.wav...


In [ ]:
def show_pydub_stats(filename):
   """Returns different audio attributes related to an audio file."""
   #creating audiosegment instance
   audio_segment = AudioSegment.from_file(filename)

   #print the pydub stats
   print(f'Channels: {audio_segment.channels}')
   print(f'Sample width: {audio_segment.sample_width}')
   print(f'Frame rate (sample rate): {audio_segment.frame_rate}')
   print(f'Frame width: {audio_segment.frame_width}')
   print(f'Length (ms): {len(audio_segment)}')

   return audio_segment

In [ ]:
#testing show pydub stats function
call1_audio_segment = show_pydub_stats(call_1_wav)

Channels: 2
Sample width: 2
Frame rate (sample rate): 32000
Frame width: 4
Length (ms): 54888


In [ ]:
def transcribe_audio(filename):
  """Takes a .wav format audio file and transcribes it to text."""
  #instantiating the class
  recognizer = sr.Recognizer()
  #importing the audio and converting to audio data
  audio_file = sr.AudioFile(filename)
  with audio_file as source:
    audio_data = recognizer.record(source)

  #transcribing the audio
  transcribed = recognizer.recognize_google(audio_data)
  return transcribed

In [ ]:
#testing the transcriber function
transcribed_call1 = transcribe_audio(call_1_wav)

The call contains 2 channels beeig channel 2 the clients speech, therefore we are going to split for only the clients part

In [ ]:
#extracting the clients part
call_1_split = call1_audio_segment.split_to_mono()
channel2 = 'call1_channel2.wav'
call_1_split[1].export(channel2, format = 'wav')
call1_channel_2 = '/content/'+channel2

The following step is to conduct a sentiment analysis with the customer call

In [ ]:
#importing the second call to apply sentiment analysis
call2 = '/content/drive/MyDrive/dataset/call2.wav'
transcribed_call2 = transcribe_audio(call2)

In [ ]:
!pip install nltk

The Audio files will be transcribed to text files, after that is completed a sentiment analysis will be conducted to infer what the client is feeling and how to address the call

In [ ]:
#importing the sentiment analysis packages
import nltk
nltk.download('punkt')
nltk.download('vader_lexicon')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package vader_lexicon to /root/nltk_data...


True

In [ ]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

In [ ]:
#instantiating the analyzer
sent = SentimentIntensityAnalyzer()
sentiment_score = sent.polarity_scores(transcribed_call2)
sentiment_score

{'neg': 0.018, 'neu': 0.888, 'pos': 0.094, 'compound': 0.778}

After the analysis is completed it's possible to conclude that the sentiment of the client is rather positive given the compund metric is 0.778 in a scale from -1 to 1. Nonetheless is important to note that the neutrality aspect is the biggest part.